runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -10412,227 +10412,732 @@ static TR::Register *inlineIsAssignableFrom(TR::Node *node, TR::CodeGenerator *c
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vadduwm, node, low4Reg, low4Reg, vtmp1Reg);
+
+   // move result to hashReg
+   generateTrg1Src2ImmInstruction(cg, TR::InstOpCode::vsldoi, node, vtmp1Reg, low4Reg, vconstant0Reg, 8);
+   generateTrg1Src1Instruction(cg, TR::InstOpCode::mfvsrwz, node, hashReg, vtmp1Reg);
+
+   // serialLabel:
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, serialLabel);
+
+   if (elementType == TR::Int32) // for Int32 there could only be 3 items at most
+      {
+      generateLabelInstruction(cg, TR::InstOpCode::b, node, special2Label);
+      }
+   else
+      {
+      // we have specialised code for cases with 3 or fewer elements
+      generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::cmpi4, node, condReg, vendReg, 3*elementSize);
+      generateConditionalBranchInstruction(cg, TR::InstOpCode::ble, node, special2Label, condReg);
+
+      const int unrollFactor = 4;
+      // vendReg = endReg - [(unrollFactor - 1) * elementSize]
+      generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addi, node, vendReg, endReg,
+         (unrollFactor-1) * elementSize * -1);
+
+      // --- unrolled loop
+      generateLabelInstruction(cg, TR::InstOpCode::label, node, serialUnrollLabel);
+      // temp = hash << 5
+      // hash = temp - hash
+      // temp = v[i]
+      // hash = hash + temp
+      generateTrg1Src2Instruction(cg, TR::InstOpCode::cmp8, node, condReg, valueReg, vendReg);
+      generateConditionalBranchInstruction(cg, TR::InstOpCode::bge, node, serialLoopLabel, condReg);
+      for (int i = 0; i < unrollFactor; i++)
+         {
+         generateTrg1Src1Imm2Instruction(cg, TR::InstOpCode::rlwinm, node, tempReg, hashReg, 5, 0xFFFFFFFFFFFFFFE0);
+         generateTrg1Src2Instruction(cg, TR::InstOpCode::subf, node, hashReg, hashReg, tempReg);
+         switch (elementType)
+            {
+            case TR::Int8:
+               generateTrg1MemInstruction(cg, TR::InstOpCode::lbz, node, tempReg,
+                     TR::MemoryReference::createWithDisplacement(cg, valueReg, i, 1));
+               if (isSigned)
+                  {
+                  generateTrg1Src1Instruction(cg, TR::InstOpCode::extsb, node, tempReg, tempReg);
+                  }
+               break;
+            case TR::Int16:
+               generateTrg1MemInstruction(cg, isSigned ? TR::InstOpCode::lha : TR::InstOpCode::lhz, node, tempReg,
+                     TR::MemoryReference::createWithDisplacement(cg, valueReg, i*2, 2));
+               break;
+            default:
+               TR_ASSERT_FATAL(false, "Unsupported hashCodeHelper elementType");
+            }
+         generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, hashReg, hashReg, tempReg);
+         }
+      generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addi, node, valueReg, valueReg, elementSize * unrollFactor);
+      generateLabelInstruction(cg, TR::InstOpCode::b, node, serialUnrollLabel);
+
```

Comment:
after this bdnz (expected to change), it is always the case that you still have at most 3 elements remaining (regardless the elementType). it is perfect to fall through to special2Label, isn't it?  get rid of this single loop completely.